contracts/Allocator.sol

```diff
@@ -303,47 +329,66 @@ contract Allocator is Permissions, IERC777Recipient {
      */
     function getTimeOfNextVest(address beneficiary) external view returns (uint) {
         ITimeHelpers timeHelpers = ITimeHelpers(contractManager.getContract("TimeHelpers"));
-        uint256 date = now;
+
         Beneficiary memory beneficiaryPlan = _beneficiaries[beneficiary];
         Plan memory planParams = _plans[beneficiaryPlan.planId - 1];
-        uint256 lockupDate = timeHelpers.addMonths(beneficiaryPlan.startMonth, planParams.vestingCliff);
-        if (date < lockupDate) {
-            return lockupDate;
+
+        uint256 firstVestingMonth = beneficiaryPlan.startMonth.add(planParams.vestingCliff);
+        uint256 lockupEndTimestamp = timeHelpers.monthToTimestamp(firstVestingMonth);
+        if (now < lockupEndTimestamp) {
+            return lockupEndTimestamp;
         }
-        uint256 dateTime = _getTimePointInCorrectPeriod(date, planParams.vestingStepTimeUnit);
-        uint256 lockupTime = _getTimePointInCorrectPeriod(
-            timeHelpers.addMonths(beneficiaryPlan.startMonth, planParams.vestingCliff),
-            planParams.vestingStepTimeUnit
-        );
-        uint256 finishTime = _getTimePointInCorrectPeriod(
-            timeHelpers.addMonths(beneficiaryPlan.startMonth, planParams.totalVestingDuration),
-            planParams.vestingStepTimeUnit
+        require(
+            now < timeHelpers.monthToTimestamp(beneficiaryPlan.startMonth.add(planParams.totalVestingDuration)),
+            "Vesting is over"
         );
-        uint256 numberOfDonePayments = dateTime.sub(lockupTime).div(planParams.vestingStep);
-        uint256 numberOfAllPayments = finishTime.sub(lockupTime).div(planParams.vestingStep);
-        if (numberOfAllPayments <= numberOfDonePayments + 1) {
-            return timeHelpers.addMonths(
-                beneficiaryPlan.startMonth,
-                planParams.totalVestingDuration
+        require(beneficiaryPlan.status != BeneficiaryStatus.TERMINATED, "Vesting was stopped");
+        
+        uint256 currentMonth = timeHelpers.getCurrentMonth();
+        if (planParams.vestingIntervalTimeUnit == TimeUnit.DAY) {
+            // TODO: it may be simplified if TimeHelpers contract in skale-manager is updated
+            uint daysPassedBeforeCurrentMonth = _daysBetweenMonths(firstVestingMonth, currentMonth);
+            uint256 currentMonthBeginningTimestamp = timeHelpers.monthToTimestamp(currentMonth);
+            uint256 daysPassedInCurrentMonth = now.sub(currentMonthBeginningTimestamp).div(_SECONDS_PER_DAY);
+            uint256 daysPassedBeforeNextVest = _calculateNextVestingStep(
+                daysPassedBeforeCurrentMonth.add(daysPassedInCurrentMonth),
+                planParams.vestingInterval
             );
-        }
-        uint256 nextPayment = finishTime
-            .sub(
-                planParams.vestingStep.mul(numberOfAllPayments.sub(numberOfDonePayments + 1))
+            return currentMonthBeginningTimestamp.add(
```

Comment:
Today is 46s day from the start, vestingInterval is 2 - so daysPassedBeforeNextVest = 48
And return would be start of 48s days
Do we vest in the begining of the day?

Reply:
We already have done so next will be less than in 2 days.

Reply:
I mean vesting was at midnight between day 45 and day 46.

Reply:
Good!